builtin/providers/aws/resource_aws_security_group.go

```diff
@@ -675,34 +678,45 @@ func resourceAwsSecurityGroupUpdateRules(
 	return nil
 }
 
+func updateSecurityGroupCache(conn *ec2.EC2, cache *Cache) error {
+	mtKey := "security_groups_cache"
+	awsMutexKV.Lock(mtKey)
+	if cache.SecurityGroups == nil {
+		req := &ec2.DescribeSecurityGroupsInput{}
+		resp, err := conn.DescribeSecurityGroups(req)
+
+		if err != nil {
+			log.Printf("Error on SGStateRefresh: %s", err)
+			return err
+		}
+
+		cache.SecurityGroups = resp
+	}
+	awsMutexKV.Unlock(mtKey)
```

Comment:
I think we want this immediately after the `Lock`, with `defer`:

```
awsMutexKV.Lock(mtKey)
defer awsMutexKV.Unlock(mtKey)
```

Otherwise don't we risk never unlocking, if we `return` in the error case above?